daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala

```diff
@@ -15,38 +14,29 @@ object Relation {
   // as the same
   // this fits our purposes for the moment
   /** A Relation. */
-  type Relation[A, B] = immutable.Map[A, Set[B]]
-
-  object Relation {
+  @deprecated("2.4.0", "use com.daml.lf.Relation")
+  type Relation[A, B] = data.Relation[A, B] // alias for Map[A, Set[B]]
+  @deprecated("2.4.0", "use com.daml.lf.Relation")
+  val Relation = this
 
-    def merge[A, B](r: Relation[A, B], pair: (A, Set[B])): Relation[A, B] =
-      r.updated(pair._1, r.getOrElse(pair._1, Set.empty[B]).union(pair._2))
+  def empty[A, B]: data.Relation[A, B] = Map.empty
 
-    def from[A, B](pairs: Iterable[(A, Set[B])]): Relation[A, B] =
-      pairs.foldLeft(immutable.Map.empty[A, Set[B]])(merge)
+  def update[A, B](r: data.Relation[A, B], a: A, b: B): data.Relation[A, B] =
+    r.updated(a, get(r, a) + b)
 
-    def union[A, B](r1: Relation[A, B], r2: Relation[A, B]): Relation[A, B] =
-      r2.foldLeft(r1)(merge)
+  def from[A, B](pairs: Iterable[(A, B)]): data.Relation[A, B] =
+    pairs.foldLeft(empty[A, B]) { case (acc, (a, b)) => update(acc, a, b) }
```

Comment:
Seems more sensical that from is based on tuple `(A, B)` instead of `(A, Set[B])`. 
I will use this is the [PackageInfo](https://github.com/digital-asset/daml/pull/14409) utility.